docs/proposals/caching.adoc

```diff
@@ -0,0 +1,77 @@
+
+With sufficient caching space available: 
+
+* A global cache directory where the parent directory is indicated by parameter `global-cache-parent`. Cleanup will only spare directories with in it so that build tasks must keep stuff in a subdirectory. For example in `$(global-cache-parent)/go-modules/`. 
+
+* A pipeline cache directory as indicated by parameter `pipeline-cache-dir`.
+
+In the initial implementation the parameters will be set to `/.cache/` and `/.cache-p/<pipeline-name>/` if space permits and be empty if there is not sufficient space available.
+
+While initially the caches will be on the same PVC as the checkout PVC, this should not be relied on, for example by using hard links between the checkout dir and the cache area. In the future we may want to keep different PVC for these areas. For example the global cache parent could be one day per project and perhaps even one per project for each build technology such as go, node. +
+Also in a future version the `pipeline-cache-dir` may be on a different PVC as the checkout PVC for example to enable faster cleanup of the pristine checkout dir by simply recreating it or not even having it on a PVC. 
+
+Files in the cache locations would persist between builds unless they are removed by the cleanup mechanism. 
+
+The build scripts are adjusted to take advantage of the cache parameters above. In addition each build task will support a new parameters `global-caching-mb` and `pipeline-caching-mb`. These defaults to 0 to indicate the task is not using caching. Otherwise it specify the minimum required cache space in MB.    
+
+All build scripts are adjusted to log
+
+- the time longer running commands take. 
+
+Furthermore the available disks pace is be logged as well (not sure where this should be done at the moment)
+
+=== ods-start and cleanup
+
+Build tasks can rely on that cache cleanup does not occur while they are running as well as that these would not be partially cleaned up. 
+
+On the other hand build Tasks cannot rely on that a cache remains populated from a previous build.
+
+The cleanup strategy described here is subject to change even in patch versions and specific details must not be relied on.
+
+Pipeline cleanup happens during `ods-start`, so that no other build can run concurrently.
+
+Before cache cleanup `ods-start` cleans up the prior PVC checkout dir, which means everything except `/.cache/<some-dir>` or `/.cache-p/<some-dir>`. In particular to prevent having files at the top levels regular files or links below `/.cache/` or `/.cache-p/` are deleted.  
+
+`ods-start` then cleans up separately the global and pipeline cache to fulfill the declared requirements of all build tasks in the odl.yaml. If disk space remains too low in a particular cache area a warning message is logged and the respective parameter `global-cache-parent` and/or `pipeline-cache-dir` will be empty. 
```

Comment:
> cleans up separately the global and pipeline cache

I think while reading through this I am actually confused what the difference is and why we need it. Could you clarify your proposal regarding this?

Reply:
The pipeline cache enable to reuse installs of dependencies in subsequent builds similar to how happens on your local machine. 
For example `npm install` will install its dependencies in `node_modules`. A subsequent build will verify that the installs are consistent with the latest `package*.json` version (which is nearly instantaneous).  Before installing a new dependency npm will always download these to a local cache first. This cache is by default  at `~/.npm`. One can reconfigure this cache to be in the global cache and would speed up the build, however the performance gain probably does not outweigh the extra space requirement in my opinion. See also https://github.com/opendevstack/ods-pipeline/discussions/411#discussion-3805693 for some impressions of performance gains.
The global cache in contrast is shared by multiple build pipelines and thus not meant to share installations of dependencies but only the dependencies itself.